src/mito2/src/manifest/checkpointer.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::Debug;
+use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+use std::sync::Arc;
+
+use common_telemetry::{error, info};
+use store_api::manifest::{ManifestVersion, MIN_VERSION};
+use store_api::storage::RegionId;
+
+use crate::manifest::action::{RegionCheckpoint, RegionManifest};
+use crate::manifest::manager::RegionManifestOptions;
+use crate::manifest::storage::ManifestObjectStore;
+use crate::metrics::MANIFEST_OP_ELAPSED;
+
+/// [`Checkpointer`] is responsible for doing checkpoint for a region, in an asynchronous way.
+#[derive(Debug)]
+pub(crate) struct Checkpointer {
+    manifest_options: RegionManifestOptions,
+    is_doing_checkpoint: Arc<AtomicBool>,
```

Comment:
We should put `is_doing_checkpoint` into the inner struct which avoids allocating an `Arc` for it.